File: ADBKit/Sources/ADBKit/Features/WindowEffects.swift
import Foundation

/// Pure math behind the translucent-window appearance: the Settings slider
/// ranges, when the effect engages, and the value each layer derives from the
/// stored opacity / blur / grain amounts. UI-free so the layering rules are
/// unit-tested.
public enum WindowEffects {
    /// Opacity slider floor — near-invisible but never fully gone, so the
    /// window stays findable over any desktop.
    public static let minimumOpacity = 0.1

    /// The opacity slider's span. 1.0 is fully opaque (the effect is off).
    public static let opacityRange: ClosedRange<Double> = minimumOpacity...1.0

    /// What a full blur slider maps to, in window-server blur radius points.
    /// Past ~40 the extra radius costs compositing time without looking any
    /// softer.
    public static let maximumBlurRadius = 40.0

    /// What a full grain slider maps to, as the noise film's alpha. Beyond
    /// this the speckle starts eating text contrast.
    public static let maximumGrainAlpha = 0.2

    /// Stored values from older builds or hand-edited defaults may be out of
    /// range — everything downstream reads through this.
    public static func clamped(_ opacity: Double) -> Double {
        guard opacity.isFinite else { return 1.0 }
        return min(max(opacity, minimumOpacity), 1.0)
    }

    /// A 0…1 slider amount (blur, grain), pinned; non-finite reads as 0.
    public static func clampedAmount(_ amount: Double) -> Double {
        guard amount.isFinite else { return 0 }
        return min(max(amount, 0), 1)
    }

    /// The effect engages only below full opacity, so 1.0 keeps the exact
    /// pre-feature rendering: opaque window, zero blur, no grain film.
    public static func isTranslucent(_ opacity: Double) -> Bool {
        clamped(opacity) < 0.999
    }

    /// Lifted surfaces (cards, bars, the sidebar) STACK on the root wash, so
    /// they carry only the contrast step: an alpha sized so root + surface
    /// composites to roughly the root plus 0.15, instead of compounding to
    /// near-solid. Opaque whenever the window is.
    public static func cardAlpha(root opacity: Double) -> Double {
        let root = clamped(opacity)
        guard isTranslucent(root) else { return 1.0 }
        return min(0.15 / max(1.0 - root, 0.15), 1.0)
    }

    /// The window-server blur radius for a slider amount — zero whenever the
    /// window is opaque (nothing shows through to blur).
    public static func blurRadius(amount: Double, root opacity: Double) -> Int {
        guard isTranslucent(opacity) else { return 0 }
        return Int((clampedAmount(amount) * maximumBlurRadius).rounded())
    }

    /// The grain film's alpha for a slider amount. Deliberately independent
    /// of the window opacity: the film reads as texture over a solid window
    /// just as well as over glass. (Blur, by contrast, needs something
    /// showing through, so it stays opacity-gated above.)
    public static func grainOpacity(amount: Double) -> Double {
        clampedAmount(amount) * maximumGrainAlpha
    }
}

File: desktop/src/lib/window-effects.ts
/**
 * The translucent-window appearance — ported from ADBKit's `WindowEffects`.
 *
 * Pure, and deliberately the same arithmetic as the Mac's rather than a
 * lookalike: the two apps have to turn the same stored opacity into the same
 * glass, or a window set up on one machine looks wrong on the other.
 *
 * One thing does *not* port, and it is named here because it is a real
 * difference rather than an omission: the Mac's Blur is a slider, driving a
 * window-server blur radius. No other platform exposes a radius — Windows has
 * Acrylic and Mica, both on-or-off, and Linux leaves window blur entirely to
 * the compositor — so `blurRadius` has no counterpart and the desktop's Blur is
 * a switch. See `docs/desktop-parity.md`.
 */

/** Opacity floor — near-invisible but never gone, so the window stays findable. */
export const MINIMUM_OPACITY = 0.1

/** Full grain maps to this alpha; past it the speckle eats text contrast. */
export const MAXIMUM_GRAIN_ALPHA = 0.2

/** Stored values can be out of range; everything downstream reads through this. */
export function clampedOpacity(opacity: number): number {
  if (!Number.isFinite(opacity)) return 1
  return Math.min(Math.max(opacity, MINIMUM_OPACITY), 1)
}

/** A 0…1 slider amount, pinned; non-finite reads as 0. */
export function clampedAmount(amount: number): number {
  if (!Number.isFinite(amount)) return 0
  return Math.min(Math.max(amount, 0), 1)
}

/**
 * The effect engages only below full opacity, so 100% keeps the exact
 * pre-feature rendering: opaque window, no blur, no grain film.
 */
export function isTranslucent(opacity: number): boolean {
  return clampedOpacity(opacity) < 0.999
}

/**
 * Lifted surfaces — the sidebar, cards, bars — STACK on the root wash, so they
 * carry only the contrast step: an alpha sized so root + surface composites to
 * roughly the root plus 0.15 instead of compounding to near-solid.
 */
export function cardAlpha(rootOpacity: number): number {
  const root = clampedOpacity(rootOpacity)
  if (!isTranslucent(root)) return 1
  return Math.min(0.15 / Math.max(1 - root, 0.15), 1)
}

/**
 * The grain film's alpha. Independent of the window opacity on purpose: the
 * film reads as texture over a solid window just as well as over glass.
 */
export function grainOpacity(amount: number): number {
  return clampedAmount(amount) * MAXIMUM_GRAIN_ALPHA
}

/** The percent readout beside a slider — the Mac's rounding, so they agree. */
export function percentLabel(value: number): string {
  return `${String(Math.round(value * 100))}%`
}

/**
 * A `#rrggbb` token re-expressed at an alpha.
 *
 * The palette tokens are hex because that is what the asset catalog holds;
 * translucency needs the same colours with an alpha channel, and rewriting the
 * palette in `rgb()` throughout would make every token harder to compare with
 * the Mac's. Anything unparseable is handed back untouched — a token that is
 * already a function or a named colour should not be corrupted into one.
 */
export function withAlpha(hex: string, alpha: number): string {
  const parsed = /^#([0-9a-f]{6})$/iu.exec(hex.trim())
  if (parsed === null) return hex
  const value = Number.parseInt(parsed[1] ?? "", 16)
  const red = (value >> 16) & 0xff
  const green = (value >> 8) & 0xff
  const blue = value & 0xff
  if (alpha >= 0.999) return hex
  // Three decimals: enough that a 1% slider step is a distinct colour, short
  // enough that the inspected value is readable.
  const rounded = Math.round(clampedAmount(alpha) * 1000) / 1000
  return `rgb(${String(red)} ${String(green)} ${String(blue)} / ${String(rounded)})`
}
